File: phi_plugin/model/getFile.ts
import fs from 'node:fs';
import csv from 'csvtojson';
import path from 'node:path';
import YAML from 'yaml';
import { logger } from '../components/Logger';



export default class getFile {

    /**
     * 读取文件
     * @param filePath 完整路径
     * @param style 强制设置文件格式
     * @param variables 需要替换的变量对象（例如：{ username: 'John' }）
     */
    static FileReader(filePath: string, style: string | 'JSON' | 'YAML' | 'TXT' = undefined, variables: Object = {}) {
        try {
            if (!fs.existsSync(filePath)) { return false }
            
            if (!style) {
                style = path.extname(filePath).toUpperCase().replace('.', '');
            }

            let fileContent = '';
            switch (style) {
                case 'JSON': {
                    fileContent = fs.readFileSync(filePath, 'utf8');
                    const parsedContent = JSON.parse(fileContent);
                    return this.replaceVariables(parsedContent, variables); // 替换占位符
                }
                case 'YAML': {
                    fileContent = fs.readFileSync(filePath, 'utf8');
                    const parsedContent = YAML.parse(fileContent);
                    return this.replaceVariables(parsedContent, variables); // 替换占位符
                }
                case 'TXT': {
                    fileContent = fs.readFileSync(filePath, 'utf8');
                    return this.replaceVariables(fileContent, variables); // 替换占位符
                }
                default: {
                    fileContent = fs.readFileSync(filePath, 'utf8');
                    return this.replaceVariables(fileContent, variables); // 替换占位符
                }
            }
        } catch (error) {
            logger.warn(`[phi-plugin][${filePath}] 读取失败`);
            logger.warn(error);
            return false;
        }
    }

    /**
     * 替换文本中的占位符
     * @param content 内容
     * @param variables 需要替换的变量对象
     */
    static replaceVariables(content: any, variables: Object) {
        // 如果没有传入 variables，直接返回原始内容
        if (Object.keys(variables).length === 0) {
            return content;
        }

        if (typeof content === 'string') {
            // 如果是字符串，替换占位符
            return content.replace(/{(\w+)}/g, (_, key) => variables[key] || `{${key}}`);
        }

        if (typeof content === 'object') {
            // 如果是对象，递归替换
            for (let key in content) {
                content[key] = this.replaceVariables(content[key], variables);
            }
        }

        return content;
    }

    static async csvReader(filePath: string) {
        return await csv().fromFile(filePath)
    }


    /**
     * 存储文件
     * @param {string} filepath 文件名，含后缀
     * @param {any} data 目标数据
     * @param {string|'TXT'|'JSON'|'YAML'} [style=undefined] 强制指定保存格式
     */
    static SetFile(filepath: string, data: any, style: string | 'JSON' | 'YAML' | 'TXT' = undefined) {
        try {
            const fatherPath = path.dirname(filepath)
            const fileName = path.basename(filepath)
            // console.info(filepath, fatherPath, fileName)
            if (!fs.existsSync(fatherPath)) {
                // 递归创建目录
                fs.mkdirSync(fatherPath, { recursive: true });
            }
            if (!style) {
                style = path.extname(filepath).toUpperCase().replace('.', '')
            }
            switch (style) {
                case 'JSON': {
                    fs.writeFileSync(filepath, JSON.stringify(data), 'utf8')
                    break
                }
                case 'YAML': {
                    fs.writeFileSync(filepath, YAML.stringify(data), 'utf8')
                    break
                }
                case 'TXT': {
                    fs.writeFileSync(filepath, data, 'utf8')
                }
                default: {
                    // Logger.error(`[phi-plugin][Set]不支持的文件格式`, style, filepath)
                    fs.writeFileSync(filepath, data, 'utf8')
                    break
                }
            }
            return true
        } catch (error) {
            // console.info(error)
            logger.warn(`[phi-plugin]写入文件 ${filepath} 时遇到错误`)
            logger.warn(error)
            return false
        }
    }


    static DelFile(path: string) {
        try {
            if (!fs.existsSync(`${path}`)) { return false }
            fs.unlink(`${path}`, (err) => {
                if (err) throw err
            })
            return true
        } catch (error) {
            logger.warn(`[phi-plugin][${path}] 删除失败`)
            logger.warn(error)
            return false
        }
    }

}

File: phi_plugin/model/getFile.py
import csv
from io import StringIO
from pathlib import Path
import re
from typing import Any

import aiofiles
from ruamel.yaml import YAML
import ujson

from ..components.Logger import logger


async def csv_write(file_path: str | Path, data: list[dict[str, Any]]):
    """
    data: list[dict]
    ```
    [
        {
            "id": "1",
            "name": "张三",
            "age": "18"
        },
        {
            "id": "2",
            "name": "李四",
            "age": "19"
        }
    ]
    """
    fieldnames = data[0].keys()

    async with aiofiles.open(file_path, mode="w", encoding="utf-8", newline="") as f:
        # 先写入表头
        csv.DictWriter(f, fieldnames=fieldnames)
        await f.write(",".join(fieldnames) + "\n")

        # 逐行写入数据
        for row in data:
            line = ",".join([str(row.get(field, "")) for field in fieldnames]) + "\n"
            await f.write(line)


async def csv_read(file_path: str | Path) -> list[dict[str, Any]]:
    """
    return list[dict]
    ```
    [
        {
            "id": 1,
            "name": "n1",
            "age": 114
        },
        {
            "id": 2,
            "name": "n2",
            "age": 39
        }
    ]
    """
    async with aiofiles.open(file_path, encoding="utf-8") as f:
        content = await f.read()

    # 使用 StringIO 模拟文件对象供 csv 模块使用
    f_obj = StringIO(content)
    reader = csv.DictReader(f_obj)
    return [dict(row) for row in reader]


class getFile:
    @staticmethod
    def FileReader(
        filePath: str | Path, style: str | None = None, variables: dict | None = None
    ):
        """
        同步读取文件

        :param filePath: 完整路径
        :param style: 强制设置文件格式['JSON' , 'YAML' , 'TXT']
        :param variables: 需要替换的变量对象 例如：{ username: 'John' }
        """
        if variables is None:
            variables = {}
        if isinstance(filePath, str):
            filePath = Path(filePath)
        try:
            if not filePath.exists():
                return False
            if not style:
                style = filePath.suffix.upper().replace(".", "")
            match style:
                case "JSON":
                    with open(filePath, encoding="utf-8") as f:
                        return getFile.replaceVariables(
                            ujson.loads(f.read()), variables
                        )
                case "YAML":
                    with open(filePath, encoding="utf-8") as f:
                        return getFile.replaceVariables(
                            YAML().load(f.read()), variables
                        )
                case _:
                    with open(filePath, encoding="utf-8") as f:
                        return getFile.replaceVariables(f.read(), variables)
        except Exception as e:
            logger.error(f"[{filePath}] 读取失败", "phi-plugin", e=e)
            return None

    @staticmethod
    def replaceVariables(content: Any, variables: dict):
        """
        替换文本中的占位符

        :param content: 内容
        :param variables: 需要替换的变量字典
        """
        if not variables:
            return content

        if isinstance(content, str):
            return re.sub(
                r"\{(\w+)\}",
                lambda match: variables.get(match.group(1), f"{{{match.group(1)}}}"),
                content,
            )
        if isinstance(content, dict):
            for key, value in content.items():
                content[key] = getFile.replaceVariables(value, variables)
            return content

        return content

    @staticmethod
    async def csvReader(filePath: str | Path):
        return await csv_read(filePath)

    @staticmethod
    def SetFile(filePath: str | Path, data: Any, style: str | None = None):
        """
        储存文件

        :param filePath: 文件路径
        :param data: 目标数据
        :param style: 强制设置文件格式['JSON' , 'YAML' , 'TXT']
        """
        if isinstance(filePath, str):
            filePath = Path(filePath)
        try:
            fatherPath = filePath.parent
            if not fatherPath.exists():
                fatherPath.mkdir(parents=True)
            if not style:
                style = filePath.suffix.upper().replace(".", "")
            match style:
                case "JSON":
                    with open(filePath, "w", encoding="utf-8") as f:
                        f.write(ujson.dumps(data))
                case "YAML":
                    with open(filePath, "w", encoding="utf-8") as f:
                        f.write(YAML().dump(data))
                case _:
                    with open(filePath, "w", encoding="utf-8") as f:
                        f.write(data)
            return True
        except Exception as e:
            logger.error(f"写入文件 {filePath} 失败", "phi-plugin", e=e)
            return False

    @staticmethod
    def DelFile(path: str | Path):
        if isinstance(path, str):
            path = Path(path)
        try:
            if not path.exists():
                return False
            path.unlink()
            return True
        except Exception as e:
            logger.error(f"[{path}] 删除失败", "phi-plugin", e=e)
            return False
